specs/capella/beacon-chain.md

```diff
@@ -0,0 +1,310 @@
+    block_number: uint64
+    gas_limit: uint64
+    gas_used: uint64
+    timestamp: uint64
+    extra_data: ByteList[MAX_EXTRA_DATA_BYTES]
+    base_fee_per_gas: uint256
+    # Extra payload fields
+    block_hash: Hash32  # Hash of execution block
+    transactions_root: Root
+    withdrawal_transactions_root: Root  # [New in Capella]
+```
+
+### New containers
+
+#### `WithdrawalTransaction`
+
+New EIP-2718 transaction type, with the format being the single byte `TX_TYPE_WITHDRAWAL`
+followed by an SSZ encoding of the `WithdrawalTransaction` container comprising the transaction contents.
+
+*Note*: This container is used for both a special TX that goes into an `ExecutionPayload`
+as well as in the `BeaconState`'s `withdrawal_receipts` queue.
+
+```python
+class WithdrawalTransaction(Container):
+    index: WithdrawalIndex
+    address: ExecutionAddress
+    amount: Gwei
+```
+
+## Helpers
+
+### Beacon state mutators
+
+#### `withdraw`
+
+```python
+def withdraw(state: BeaconState, index: ValidatorIndex, amount: Gwei) -> None:
+    # Decrease the validator's balance
+    decrease_balance(state, index, amount)
+    # Create a corresponding withdrawal receipt
+    receipt = WithdrawalTransaction(
+        index=state.withdrawal_index,
+        address=state.validators[index].withdrawal_credentials[12:],
+        amount=amount,
+    )
+    state.withdrawal_index = WithdrawalIndex(state.withdrawal_index + 1)
+    state.withdrawal_receipts.append(receipt)
+```
+
+### Predicates
+
+#### `is_fully_withdrawable_validator`
+
+```python
+def is_fully_withdrawable_validator(validator: Validator, epoch: Epoch) -> bool:
+    """
+    Check if ``validator`` is fully withdrawable.
+    """
+    is_eth1_withdrawal_prefix = validator.withdrawal_credentials[0:1] == ETH1_ADDRESS_WITHDRAWAL_PREFIX
+    return is_eth1_withdrawal_prefix and validator.withdrawable_epoch <= epoch < validator.withdrawn_epoch
+```
+
+## Beacon chain state transition function
+
+### Epoch processing
+
+```python
+def process_epoch(state: BeaconState) -> None:
+    process_justification_and_finalization(state)
+    process_inactivity_updates(state)
+    process_rewards_and_penalties(state)
+    process_registry_updates(state)
+    process_slashings(state)
+    process_eth1_data_reset(state)
+    process_effective_balance_updates(state)
+    process_slashings_reset(state)
+    process_randao_mixes_reset(state)
+    process_historical_roots_update(state)
+    process_participation_flag_updates(state)
+    process_sync_committee_updates(state)
+    process_full_withdrawals(state)  # [New in Capella]
+```
+
+#### Withdrawals
+
+*Note*: The function `process_full_withdrawals` is new.
+
+```python
+def process_full_withdrawals(state: BeaconState) -> None:
+    current_epoch = get_current_epoch(state)
+    for index, validator in enumerate(state.validators):
+        if is_fully_withdrawable_validator(validator, current_epoch):
+            # TODO, consider the zero-balance case
```

Comment:
if the validator is `fully_withdrawable` and has `0` balance, then one option is to skip the call to `withdraw` (to not clutter the receipts queue) although we still want to update `validator.withdrawn_epoch`

if we went this route, just need to make sure no important accounting (e.g. to signal validator record re-useability) is done in the function `withdraw`

Reply:
I think we may have zero balance case handled inside of `withdraw` function. This makes the function well isolated and responsible for entire withdrawal flow which prevent potential issues caused by forgetting anything related. Also,  `withdraw_epoch` should be set inside of this function.

Reply:
This function will also be used for partial withdrawals (e.g. skimming off balances above 32ETH during block proposal or some othr method). That's why it does *not* update `withdrawal_epoch` and why the balance is passed in (insteads of an assumed max balance withdrawal)

Reply:
For the sake of single responsibility principle, I would make a separate function for partial withdrawals if this does make sense.

Reply:
For the sake of code reuse and testing, I'm not actually sure. We have instances of core functionality making it into helper functions and used in multipel contexts -- e.g. slash_validator, initiate_validator_exit, etc